Add some docs for .checked and .on_failure (#1301)
sorbet/sorbet: website/docs/runtime.md

checks run to determine whether the `T::Configuration` handler should be called
in the first place. This comes with an overhead, and we carefully audit and
monitor the performance of `sorbet-runtime` as a result. The overhead of these
checks are usually very small.

But in some cases, especially when calling certain methods in tight loops or
other latency-sensitive paths, the overhead of even doing the checks (regardless
of what happens on failure) is prohibitively expensive. To handle these cases,
Sorbet offers `.checked(...)` which declares in what environments a sig should
be checked:

```ruby
# (1) Runtime checks always run.
sig {params(xs: T::Array[String]).void.checked(:always)}

# (2) Runtime checks only run in "tests" (see below).
# In non-tests, this sig specifically has no runtime overhead.
sig {params(xs: T::Array[String]).void.checked(:tests)}

# (3) Never runs the runtime checks. Careful!
sig {params(xs: T::Array[String]).void.checked(:never)}
```

If `.checked(...)` is omitted on a sig, the default is `.checked(:always)`. The
default checked level can also be configured. For example:

```ruby

> T::Configuration.default_checked_level = :tests

Reviewer: I think it would be helpful to mention that this setting cannot be changed later on in the code and it must be set right after Sorbet is loaded.

runtime checks. But it's precisely these runtime checks that make it easier to
drive adoption of types in the long run.

> ## Changing the runtime behavior

Reviewer 2: Before going in depth with how to change the runtime behaviour, it may be better to give an example as why you would want to change its behaviour.

I think it also worth noting that code raising no error with Sorbet static could actually raise some with Sorbet runtime (and even in production).

Author: Good suggestion, thanks!

Author: Added blurb + snippet above.

example at Stripe, we use `.on_failure` to attach team ownership information to
a failure.

The `T::Configuration` handler we wrote branches on whether `.on_failure` was
provided, which means the logging behavior is opt in. This is nice because it
means the default behavior is still to make problems with types fail loudly and
early, rather than silently as a log. If we wanted, we could have inverted this:

```ruby
T::Configuration.call_validation_error_handler = lambda do |signature, opts|
  #                                                     ┌────┐
  if signature.on_failure && signature.on_failure[0] == :raise
    raise TypeError.new(opts[:pretty_message])
  else
    puts opts[:pretty_message]
  end
end

# ...

#                                       ┌─────────────────┐
sig {params(argv: T::Array[String]).void.on_failure(:raise)}
```

With this `T::Configuration` handler, the default is to log, and we can use
`.on_failure` to opt specific sigs into raising on failure.

For more on the various `T::Configuration` handlers, see
[Runtime Configuration](tconfiguration.md).

## `.checked`: Whether to check in the first place

> > **Careful!** Opting out of runtime checks opts out of significant safety

Reviewer 3: I think this is a little technical and vague to someone who don't already know what the consequences are. I would say something like `Opting out of runtime checks can significantly degrade the trustworthiness of type signatures. Only
disable the runtime when you're aware of the tradeoffs.` to be more visceral. Maybe also link to https://sorbet.org/docs/gradual for someone who wants to understand the trade-offs?

Reviewer 4: It might be worth noting (possibly here, possibly elsewhere) the kind of failure cases that arise from turning off checking, and in particular the possibility that silently violating a type contract in one place can cause errors to arise in a different place entirely.

> checks Sorbet relies on when making claims about a codebase statically. Only
> disable the runtime when you're aware of the tradeoffs.

In our examples above with `.on_failure`, every method call had the runtime type
checks run to determine whether the `T::Configuration` handler should be called
in the first place. This comes with an overhead, and we carefully audit and
monitor the performance of `sorbet-runtime` as a result. The overhead of these
checks are usually very small.

But in some cases, especially when calling certain methods in tight loops or
other latency-sensitive paths, the overhead of even doing the checks (regardless
of what happens on failure) is prohibitively expensive. To handle these cases,

> Sorbet `.checked(...)` which declares in what environments a sig should be

Reviewer 3: ```suggestion
Sorbet offers `.checked(...)` which declares in what environments a sig should be
```

checked:

```ruby
# (1) Runtime checks always run.
sig {params(xs: T::Array[String]).void.checked(:always)}

# (2) Runtime checks only run in "tests" (see below).

> # In non-tests, the sig is **only** has no runtime overhead.

Reviewer 3: ```suggestion
# In non-tests, **only** this sig has no runtime overhead.
```
